### PR TITLE
feat(ci) Add an optional mypy check to Sentry

### PR DESCRIPTION
**Pull request by evanh**
Add an optional mypy check to Sentry. It will run on only the files specified in
the mypy config file.

Also I assume this won't block PRs at all, but I can't tell because I don't have access to the settings for the Sentry repo.

### CODE REVIEW

File: .github/workflows/backend-typing.yml
Comment (billyvg) [inline .github/workflows/backend-typing.yml:39]: It'd be nice if we could skip the rest of the steps if no files specified in mypy were modified

Comment (evanh) [reply .github/workflows/backend-typing.yml:39]: That's a good idea. If you have suggestions of other unnecessary steps to trim from this, that would be helpful too. It seems like a lot of steps to run mypy.

Comment (evanh) [reply .github/workflows/backend-typing.yml:39]: Do you have a suggestion of how to avoid duplicating the filenames in the file-filters as well as mypy.ini? I'm trying to make typing as frictionless as possible and it would be nice if people only had to update one file.

File: .github/workflows/backend-typing.yml
Comment (armenzg) [inline .github/workflows/backend-typing.yml:6]: Do we still use `releases` branches?

File: .github/workflows/backend-typing.yml
Comment (armenzg) [inline .github/workflows/backend-typing.yml:69]: Do you need to set up all of Sentry? Or could you get by with `pip install -r requirements-dev.txt`?

File: .github/workflows/backend-typing.yml
Comment (evanh) [reply .github/workflows/backend-typing.yml:6]: I'm not sure. I copied this from the snuba test workflow.

File: .github/workflows/backend-typing.yml
Comment (evanh) [reply .github/workflows/backend-typing.yml:69]: I just need mypy, this isn't running Sentry at all. I can change this.

Comment (evanh) [reply .github/workflows/backend-typing.yml:69]: I spoke too soon. Removing the sentry env setup caused mypy to run against every file instead of just the ones specified in the mypy config. Will have to investigate further.

File: .github/workflows/backend-typing.yml
Comment (billyvg) [reply .github/workflows/backend-typing.yml:39]: Hmm I don't have any suggestions yet, let's not block on this, instead maybe add a ticket for @getsentry/productivity to look into it in the future. If this workflow runs fast enough it shouldn't be a big deal.

Comment (evanh) [reply .github/workflows/backend-typing.yml:39]: How do I add a ticket?

Comment (billyvg) [reply .github/workflows/backend-typing.yml:39]: https://github.com/getsentry/sentry/issues/new/choose :D - in `Projects` on the RHS, add it to CI/tooling

**Review by wedamija — COMMENTED** [review]
I don't have too much to add to this review, but thanks for getting this going!

File: .github/workflows/backend-typing.yml
Comment (evanh) [reply .github/workflows/backend-typing.yml]: OK I think I found a minimal way to set this up correctly.

File: .github/workflows/backend-typing.yml
Comment (armenzg) [reply .github/workflows/backend-typing.yml:6]:
```suggestion
```

File: .github/workflows/backend-typing.yml
Comment (armenzg) [inline .github/workflows/backend-typing.yml:13]:
```suggestion
    timeout-minutes: 12
```

I see it taking 7 minutes, thus, reducing it.

File: .github/workflows/backend-typing.yml
Comment (armenzg) [inline .github/workflows/backend-typing.yml:16]: Remove this if we're not going to have more than 1.

File: .github/workflows/backend-typing.yml
Comment (armenzg) [inline .github/workflows/backend-typing.yml:24]: You can remove this `with` code if you're not going to be submitting data to codecov.

File: .github/workflows/backend-typing.yml
Comment (armenzg) [reply .github/workflows/backend-typing.yml:39]: Did you get to this?

File: .github/workflows/backend-typing.yml
Comment (armenzg) [inline .github/workflows/backend-typing.yml:81]: If you're not uploading code coverage you can remove this.

**Review by armenzg — CHANGES_REQUESTED** [review]
Thank you so much for tackling this!

File: .github/workflows/backend-typing.yml
Comment (evanh) [reply .github/workflows/backend-typing.yml:39]: I didn't want to add it until I actually had something in Sentry. Seemed weird to make a feature request for a feature that didn't exist yet. I planned to open a ticket once this was approved/merged.

**Review by armenzg — APPROVED** [review]

**Pull request merged**
